### PR TITLE
Fixes #22374 - adapt to use importing from core

### PR DESCRIPTION
**Pull request by ares**
This illustrates how we could use importing from core in this plugin, https://github.com/theforeman/foreman/pull/5201 would need to be merged first

### CODE REVIEW

**Comment by xprazak2** [comment]
Any plans moving ahead with this now that [core](https://github.com/theforeman/foreman/pull/5201) part has been merged or should I?

**Comment by ares** [comment]
absolutely, just need to find a bit more time for this

**Comment by ares** [comment]
Tests are broken by core changes, here's the output of local tests run

```
ares@edna:[~/z/foreman][develop✓]$ rake test:foreman_templates
Run options: --seed 64895

# Running:

...............................................

Finished in 2.534629s, 18.5431 runs/s, 44.5825 assertions/s.
47 runs, 113 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by ares** [comment]
Latest run with, locking requires https://github.com/theforeman/foreman/pull/5313 but does not block this PR, I can squash if you prefer.

```
Run options: --seed 47161

# Running:

...............................................

Finished in 3.029199s, 15.5157 runs/s, 37.3036 assertions/s.
47 runs, 113 assertions, 0 failures, 0 errors, 0 skips
```

File: app/services/foreman_templates/template_importer.rb [hunk 1]
```diff
@@ -54,54 +51,52 @@ def parse_files!
...
+            template_type = metadata['model'].constantize
+          else
+            result_lines << "Metadata for template '#{name}' do not specify 'model', import failed"
+            next
+          end
+
+          template = template_type.import!(name, text, options)
+
+          data = {}
+          if template.template_changed?
```
Comment (xprazak2) [inline app/services/foreman_templates/template_importer.rb]: Will always be `false`. `template_type.import!` on l.77 returns newly saved template, no diffs are ever shown.

**Comment by xprazak2** [comment]
I also tested with the [locking](https://github.com/theforeman/foreman/pull/5313), encountered some unexpected behaviour. 

1) have an unlocked template
2) make change in template content
3) try to import with `'lock': 'true'`
4) you get an error that template is locked, but no changes to template content or lock are made

File: app/services/foreman_templates/template_importer.rb [hunk 2]
```diff
@@ -54,54 +51,52 @@ def parse_files!
         result_lines << 'Parsing: ' + template.gsub(/#{@dir}#{@dirname}/, '') if @verbose
 
         metadata = parse_metadata(text)
```
Comment (xprazak2) [inline app/services/foreman_templates/template_importer.rb]: Shouldn't we rather use `Template::parse_metadata`?

Comment (ares) [reply app/services/foreman_templates/template_importer.rb]: :1st_place_medal:

File: app/services/foreman_templates/template_importer.rb (hunk 1 above)
Comment (ares) [reply app/services/foreman_templates/template_importer.rb]: bummer, `import!` seems to be too high-level, we need to hook into the process of saving to get the diff before save is called, that would actually help solving the problem with locking too, I'll look into it

**Comment by ares** [comment]
> TemplateImporter#update_job_template is no longer used and can be removed, same for status_to_text.

Good catches!

> I also tested with the locking, encountered some unexpected behavior.

Did you also set `enforce: true`? Did the user have permission? Did you use rake task or API call? In general, we don't have control over bypassing the lock from this level as we don't have the instance yet. We could always pass enforce param, but then it kind of defeats the purpose. Note that user has to have permissions to lock the template - https://github.com/theforeman/foreman/blob/develop/app/models/template.rb#L182

Or was it this error? https://github.com/theforeman/foreman/blob/develop/app/models/template.rb#L193 I don't see how that would happen if we actually pass locked parameter, could you please check it bubbled there?

**Comment by ares** [comment]
Updated, it seems to work with diff now, @xprazak2 please take another look.

**Comment by xprazak2** [comment]
The diff shows now.

The locking validation error was indeed what [you suggested](https://github.com/theforeman/foreman/blob/develop/app/models/template.rb#L193). I was running as admin via API without `force`. The thing is that when you modify template text and try to lock it at the same time you have:

```ruby
actual_changes
#=> { :template => "whatever", :locked => [false, true] }

```

and the both parts of condition are true because we do not force it and `template` attribute is not in `allowed_changes`. But it will have to be fixed separately. Workaround is to do it in 2 steps - first change the template and then lock it.

I think you can squash and attach this to a [redmine issue](https://projects.theforeman.org/issues/22374)

**Comment by ares** [comment]
After the change in core, I think this should now behave as expected. I squashed commits and linked it to the redmine issue, could you please take another look @xprazak2 ?

**Comment by xprazak2** [comment]
Seems we are good here, the tests are now broken because of change in core, so this should actually fix them. Merging, thanks!

**Pull request merged**